### PR TITLE
Reporting

### LINKED ISSUE #41: Reporting class
Start with what Stadler does with their MIA attacks. 

Class MiAttackSummary (which is the child of a AttackSummary class)

attributes:

- predictions
- test_labels
- generator info
- attack info
- metrics?

methods:

- calculate metrics (starting from [here](https://github.com/alan-turing-institute/synthetic_data_release/blob/95001943e0f3b021c606c3c8ed67204839a969d2/utils/evaluation_framework.py#L12)).
- save metrics to file?

Here an instance of the class is for one given attack-generator-dataset combination. If we want to compare different attacks/generators as done in Stadler we will need a class with a higher level of abstraction and plotting  methods.

### PR DESCRIPTION
**Pull request by crangelsmith**
Fixes #41 

Reporting figure for MIA attacks.

### CODE REVIEW

File: prive/report/report.py [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+    @abstractmethod
+    def write(self, output_path):
+        """
+        Write metrics to file.
+
+        """
+        pass
+
+
+class MiAttackSummary(AttackSummary):
```
Comment (JamesJordon) [inline prive/report/report.py:25]:
```suggestion
class MIAttackSummary(AttackSummary):
```
to be consistent with other names

**Comment by crangelsmith** [comment]
@JamesJordon the results obtained from [here](https://github.com/alan-turing-institute/privacy-sdg-toolbox/blob/7b0a8a8905311eaeeccf25bbf93e14732256c90f/prive/run_groundhog.py#L56), particularly `test_labels` and `predictions` are the results of a single attack, or multiple attacks concatenated into one array?

The `MIAttackSummary` class currently is thought as the summary metrics for one given attack, so I'm trying to think how to build summaries over many 'games' of the same attack.

**Comment by JamesJordon** [comment]
> @JamesJordon the results obtained from [here](https://github.com/alan-turing-institute/privacy-sdg-toolbox/blob/7b0a8a8905311eaeeccf25bbf93e14732256c90f/prive/run_groundhog.py#L56), particularly `test_labels` and `predictions` are the results of a single attack, or multiple attacks concatenated into one array?
> 
> The `MIAttackSummary` class currently is thought as the summary metrics for one given attack, so I'm trying to think how to build summaries over many 'games' of the same attack.

The results there are for a single attack, and a single target. A big TODO right now is to work out how we want to handle attacks over multiple targets.

**Comment by crangelsmith** [comment]
Here an example of the kind of figures the report class can create for a dataset with summary statistics for different attacks.

In this example the attack summary dummy data is from 1 raw dataset (`TestData1`),  two types of attacks (`RandomGuess1`,`RandomGuess2`) and three generators (`RandomGenerator1'`,`RandomGenerator2`,`RandomGenerator3`). Basically, the reporting creates figures of the performance metrics for all different combinations of data-attack-generators.

attacksComparison_DatasetTestData1_AttackRandomGenerator1
generatorsComparison_DatasetTestData1_AttackRandomGuess2
datasetsComparison_DatasetRandomGuess2_AttackRandomGenerator2

**Comment by crangelsmith** [comment]
After our discussion last Tuesday I have added a couple of more features:

- MIAReport class has two methods, compare (configurable for dimensions to compare) and create_report which create all possible combinations (including target id as x axis and generation/attacks as points).
- Axis ranges are configurable per metric.
- Aesthetic setup.

Here examples of the latest figures:

attacksComparison_DatasetTestData1_AttackRandomGenerator1
target_idsComparison_DatasetTestData1_AttackRandomGuess2
generatorsComparison_DatasetTestData1_AttackRandomGuess2
 
datasetsComparison_DatasetRandomGuess1_AttackRandomGenerator3

**Comment by triangle-man** [comment]
Can I propose we merge this pull request?

**Pull request merged**